### ISSUE TITLE
tail sampling traces

### ISSUE THREAD

**Issue by rbtcollins** [vectordotdev/vector#16929]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

We use opentelemetry to emit traces from our clusters. The volume of traces is quite high. However most traces don't offer value.

Some vendors like (Honeycomb)[https://www.honeycomb.io/] have (tail samplers)[https://docs.honeycomb.io/manage-data-volume/refinery/] that dramatically reduce the number of traces that need to be kept to provide a holistic view of the running service.

Being able to this outside of proprietary vendor tooling would be great.

tl;dr: see low cardinality events but reduce span egress and storage 80% or more

### Attempted Solutions

I looked but couldn't see anything in the docs about tail sampling.

However from an architecture perspective I'd expect something like:

source(otel) -> tailsampler w/5GB look-back ->sink(otel to vendor GRPC endpoint)

### Version

we haven't adopted vector at this point

**Comment by zamazan4ik** [comment]
Well, it could be implemented via a simple `random` function on a transformation step with VRL. Something like this (warning - pseudocode):
```
let value = random(100);
if (value < 10)
    pass_logs_to_sink();
```

But there is no such a function in VRL yet - @jszwedko probably could help here.

**Comment by rbtcollins** [comment]
Thats not what is implied by tail sampling in the tracing domain. Head sampling can do that random based sampling on the trace id and pass non-recording spans down into the stack.

Have a look at the refinery docs for more details but the core concept is to signal boost.
For instance, you can build a tuple (operation, error-status) and then:
- buffer all spans for a trace (maybe 5-10, or possibly 10's of K of spans)
- when some threshold is reached, or the root span is received (which should have error rolled-up to the root) then making a sample decision.
- if the current sample target is higher than the sent traces, sample the trace
- if the relative cardinality for any of the tuples in the trace is lower than the average sampled cardinality, sample the trace

**Comment by jszwedko** [comment]
Thanks for opening this @rbtcollins !

We discussed this issue a bit today. It is something we think fits in the vision of Vector but is a heavy lift to add since currently Vector has no shared state between instances, which seems to be a requirement for this feature. For that reason, it's unlikely that we'll add this in the near future.

It would be easier to add a local-only tail sampling that only looks at traces received by a single Vector instance.

**Comment by srstrickland** [comment]
FWIW, even otel collector has no shared state between instances.  To get it to the "right" thing you have to be sure all spans of the same trace end up in the same instance.  The two ways I know of doing it for otel are:
1. Use kafka with a key=trace_id
2. Use a loadbalancing exporter to route data to a downstream collector based on the `trace_id` attribute

As it is, vector -> vector communication doesn't allow any attribute-based routing, but I think using kafka is a reasonable way to achieve the necessary colocation.

**Comment by mugli** [comment]
One potential downside of having no shared state in OTel collector (and routing traces based on just `trace_id` attribute with the loadbalancing exporter) is this: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33568

Traces from async messaging systems don't always share the same `trace_id`, but can use Span Links instead ([an example with Kafka](https://opentelemetry.io/docs/specs/semconv/messaging/kafka/#apache-kafka-with-quarkus-or-spring-boot-example))

Currently there's no way to ensure if a particular trace is sampled in the OTel tailsampler, all related linked traces will be sampled as well.

The above linked issue has some discussion around it. Wanted to share here, in case Vector decides to support tailsampling for trace it would be nice to avoid the same gotcha.

**Comment by pront** [comment]
There is an open feature request that might help with routing: https://github.com/vectordotdev/vector/issues/23583
